Source/Core/Core/AchievementManager.cpp

```diff
@@ -0,0 +1,114 @@
+#include "rcheevos/include/rc_api_user.h"
+#include "rcheevos/include/rc_hash.h"
+#include "rcheevos/include/rc_runtime.h"
+
+#include "Common/HttpRequest.h"
+#include "Config/AchievementSettings.h"
+#include "Core/Core.h"
+
+namespace Achievements
+{
+static rc_runtime_t runtime{};
+static bool is_runtime_initialized = false;
+static rc_api_login_response_t login_data{.response{.succeeded = 0}};
+
+namespace  // Hide from use outside this file
+{
+template <typename RcRequest, typename RcResponse>
+void Request(RcRequest rc_request, RcResponse* rc_response,
+             int (*init_request)(rc_api_request_t* request, const RcRequest* api_params),
+             int (*process_response)(RcResponse* response, const char* server_response))
+{
+  rc_api_request_t api_request;
+  Common::HttpRequest http_request;
+  init_request(&api_request, &rc_request);
+  auto http_response = http_request.Post(api_request.url, api_request.post_data);
+  if (http_response.has_value() && http_response.value().size() > 0)
+  {
+    char* response_str = (char*)calloc(http_response.value().size() + 1, 1);
+    for (int ix = 0; ix < http_response.value().size(); ix++)
+    {
+      response_str[ix] = (char)http_response.value()[ix];
+    }
+    process_response(rc_response, (const char*)response_str);
+    free(response_str);
+  }
+  rc_api_destroy_request(&api_request);
+}
+}  // namespace
+
+void Init()
+{
+  if (!is_runtime_initialized && Config::Get(Config::RA_INTEGRATION_ENABLED))
+  {
+    rc_runtime_init(&runtime);
+    is_runtime_initialized = true;
+  }
+}
+
+void Login()
+{
+  if (!Config::Get(Config::RA_INTEGRATION_ENABLED) || !is_runtime_initialized ||
+      Config::Get(Config::RA_USERNAME).empty() || Config::Get(Config::RA_API_TOKEN).empty())
+    return;
+  std::string username = Config::Get(Config::RA_USERNAME);
+  std::string api_token = Config::Get(Config::RA_API_TOKEN);
+  rc_api_login_request_t login_request = {.username = username.c_str(),
+                                          .api_token = api_token.c_str()};
+  Request<rc_api_login_request_t, rc_api_login_response_t>(
+      login_request, &login_data, rc_api_init_login_request, rc_api_process_login_response);
+  if (!login_data.response.succeeded)
+    return;
+  ResetGame();
+}
+
+std::string Login(std::string password)
+{
+  if (!Config::Get(Config::RA_INTEGRATION_ENABLED) || !is_runtime_initialized ||
+      Config::Get(Config::RA_USERNAME).empty() || password.empty())
+    return "";
+  std::string username = Config::Get(Config::RA_USERNAME);
+  rc_api_login_request_t login_request = {.username = username.c_str(),
+                                          .password = password.c_str()};
+  Request<rc_api_login_request_t, rc_api_login_response_t>(
+      login_request, &login_data, rc_api_init_login_request, rc_api_process_login_response);
+  if (!login_data.response.succeeded)
+    return "";
+  ResetGame();
+  return std::string(login_data.api_token);
+}
```

Comment:
These two functions can definitely be combined; extract the `login_request` into a parameter.

The `ResetGame()` call here also bothers me. My game shouldn't close just because I'm logging into an account. I think it would be better to just disallow logging in while a game is running if that's a problem.

Reply:
In terms of combining, I can give you a third function, possibly private, that takes a login_request parameter that both the other functions call; I'm maybe 85% certain this is what you're asking for.

As for ResetGame(), I did get feedback at one point that we might want to do something similar with hardcore, i.e. only change the state of that setting when a game isn't currently running. I do think that makes sense but my only concern is a potential race condition if Dolphin is directed to boot a game upon startup. This is designed to log in automatically on startup if credentials are saved from a previous session, but I'd want to make sure that's active _before_ the game is booted.

Reply:
> In terms of combining, I can give you a third function, possibly private, that takes a login_request parameter that both the other functions call; I'm maybe 85% certain this is what you're asking for.

Yeah that's what I mean.

> race condition

I'll grant that our threading is... not the best at times, but this seems like a strange way to handle that. If you really want to make this safe, you could pass a flag whether achievements are allowed to the Boot function or something to that effect.

Reply:
In fact, if your goal is to associate a single game session with a single achievement account, you should probably pass all the necessary account info (userid, sessionid or whatever) into the Boot function and then only use that while the game is running.

Reply:
Had a bit of time to think (could say I slept on it, really) and now I'm not entirely certain why I was concerned about this. I think what I was getting at was more a matter of, if we only allow logging in while a game is not running then I need to ensure this happens completely before the game boots and ... I think I already do that.

That said I'm not sure about putting this information into the boot function because I do want to remain logged in while a game is _not_ running. I'll look into it and consider it though.

Reply:
Also I went to go make the change you were talking about and the shared method is only one line of code, that seems a bit excessive. I already have the base method on line 25 for making RetroAchievements-standard API calls, which will get used a LOT in later work, because every API call we have (with the only exception being images) has the same init request - process response - delete response structure. But here ultimately most of the work in each login method is setting up the request object and then the only thing in common is the actual call on line 65/80, which is already calling a common Request method.